storage: migrate raft state on startup

tbg opened the pull request:
In #7310, we added (and enforced) assumptions on what was
persisted to disk, but did not ensure that older versions
were migrated properly. This in turn lead to #7389.

A migration is only necessary for ranges which had never
been truncated, and so it should be fairly rare in practice.

The migration simply writes the truncated state that would
otherwise been synthesized by the old code, which is technically
a consistency violation, but should work well in practice because
all nodes will write that state on startup when they are rebooted.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7429)

<!-- Reviewable:end -->

mberhault commented:
LGTM, but I'm not an expert

tschottdorf commented:
Yeah, we'll try this out and then hopefully @bdarnell can take a look.

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store.go, line 908 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKy3ZvXdGZ1W1nTURCv:-KKy3ZvXdGZ1W1nTURCw:-1643775137) ([raw file](https://github.com/cockroachdb/cockroach/blob/86090be5fd7478a4616157f238ccea61eae7d37c/storage/store.go#L908)):*
> ```Go
> 			}
> 			log.Warning("migration: synthesized truncated state for %+v", desc)
> 		}
> ```

We need a test for this, though it doesn't need to be onerous. Seems feasible to pull out the above into a separate method and to construct a test which writes the old state and verifies that the migration updates the state as expected. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: migrate raft state on startup

In #7310, we added (and enforced) assumptions on what was
persisted to disk, but did not ensure that older versions
were migrated properly. This in turn lead to #7389.

A migration is only necessary for ranges which had never
been truncated, and so it should be fairly rare in practice.

The migration simply writes the truncated state that would
otherwise been synthesized by the old code, which is technically
a consistency violation, but should work well in practice because
all nodes will write that state on startup when they are rebooted.

tschottdorf commented:
Updated, PTAL. Tested a Docker reference test. I initially tried to unit-test this, but it was really silly and not convincing at all.

tamird commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 4 of 4 files at r3.
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[acceptance/reference_test.go, line 34 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzryYz5WPRGozJeRaL:-KKzryYz5WPRGozJeRaM:-446655245) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/reference_test.go#L34)):*
> ```Go
> 	err := testDockerOneShot(t, "reference", []string{"/bin/bash", "-c", script})
> 	if err != nil {
> 		t.Errorf("expected success: %s", err)
> ```

heh why not just t.Error(err)

---

*[acceptance/replication_test.go, line 44 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzs5ayuOJb_82ElsTo:-KKzs5ayuOJb_82ElsTp:-1679476492) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/replication_test.go#L44)):*
> ```Go
> 		// Looks silly, but we actually start zero-node clusters in the
> 		// reference tests.
> 		t.Log("replication test is a no-op for empty cluster")
> ```

t.Skip("replication test ...") instead of this return

---

*[acceptance/util_test.go, line 214 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsDcHt0eYSSuDwMIy:-KKzsDcHt0eYSSuDwMIz:-1163759776) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L214)):*
> ```Go
> 			for ; i < 100; i++ {
> 				_, _, fun := caller.Lookup(i)
> 				if testFuncRE.MatchString(fun) {
> ```

surely i'm misreading this?

---

*[acceptance/util_test.go, line 288 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsLwESVnWQL-A2p_T:-KKzsLwESVnWQL-A2p_U:1936594020) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L288)):*
> ```Go
> )
> 
> func testDocker(t *testing.T, num int, name string, cmd []string) error {
> ```

nit: just make this int32, you're dealing with constants here

---

*[acceptance/cluster/localcluster.go, line 228 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzrJigpgwhn0d371xv:-KKzrJigpgwhn0d371xw:1647197056) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/cluster/localcluster.go#L228)):*
> ```Go
> 		return err
> 	}
> 	err = l.oneshot.Wait()
> ```

reduce scope?

---

*[acceptance/cluster/localcluster.go, line 232 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzrLXEiMjgQ33oQlNZ:-KKzrLXEiMjgQ33oQlN_:1085724451) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/cluster/localcluster.go#L232)):*
> ```Go
> 		var b bytes.Buffer
> 		l.oneshot.Logs(&b)
> 		return fmt.Errorf("%s:\n%s", err, b.String())
> ```

errors.Wrap? :trollface: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[acceptance/reference_test.go, line 34 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzryYz5WPRGozJeRaL:-KKzvtzO8_l3Q3kBBgo9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/reference_test.go#L34)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> heh why not just t.Error(err)
</details>
Done.

---

*[acceptance/replication_test.go, line 44 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzs5ayuOJb_82ElsTo:-KKztuM6ZLCJaeU0nOVU:1751540390) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/replication_test.go#L44)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> t.Skip("replication test ...") instead of this return
</details>
Nope, I don't want to skip the tests that call this. Just not test this.

---

*[acceptance/util_test.go, line 214 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsDcHt0eYSSuDwMIy:-KKzuD25GBE8OLXyWBDC:-1992998233) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> surely i'm misreading this?
</details>
Erm. Fixed.

---

*[acceptance/util_test.go, line 288 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsLwESVnWQL-A2p_T:-KKzuVLUI6fcXTj2T8Dm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L288)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: just make this int32, you're dealing with constants here
</details>
Done.

---

*[acceptance/cluster/localcluster.go, line 228 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzrJigpgwhn0d371xv:-KKzud1dY6bf2zcru-Zs:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/cluster/localcluster.go#L228)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> reduce scope?
</details>
Done.

---

*[acceptance/cluster/localcluster.go, line 232 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzrLXEiMjgQ33oQlNZ:-KKzurr6VBczUEtLA_fy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/cluster/localcluster.go#L232)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> errors.Wrap? :trollface: 
</details>
Done.

---

*[storage/store.go, line 908 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKy3ZvXdGZ1W1nTURCv:-KKzusvjAQ_RADXCuoLO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/86090be5fd7478a4616157f238ccea61eae7d37c/storage/store.go#L908)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We need a test for this, though it doesn't need to be onerous. Seems feasible to pull out the above into a separate method and to construct a test which writes the old state and verifies that the migration updates the state as expected. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[acceptance/replication_test.go, line 44 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzs5ayuOJb_82ElsTo:-KKzwBRo3qRET6wSlNFy:472301517) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/replication_test.go#L44)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Nope, I don't want to skip the tests that call this. Just not test this.
</details>
Oh, I just assumed this was a test function.

---

*[acceptance/util_test.go, line 214 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsDcHt0eYSSuDwMIy:-KKzwd5SrRsdTnQWAoFj:-1002467270) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L214)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Erm. Fixed.
</details>
i'd put the logic in here instead of continue and break but not worth another CI run

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-:-KKzwiWRzn8F2ZpzJZl_:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[acceptance/util_test.go, line 214 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsDcHt0eYSSuDwMIy:-KKzwjfUoSnDrRIOsflM:618460696) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i'd put the logic in here instead of continue and break but not worth another CI run
</details>
oh wait you need to bump the image tag, so you can make that change :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[acceptance/util_test.go, line 214 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429#-KKzsDcHt0eYSSuDwMIy:-KL--C1wtvZrxfBkXD5x:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/34b9892ea53577b2873a2e52d431616cb314c5b8/acceptance/util_test.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh wait you need to bump the image tag, so you can make that change :)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7429)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
`postgres-test` version bumped, tested again locally, merging on green.

tbg committed:
acceptance: add reference test for #7389

Also straightened out some kinks with the existing reference tests, namely that
they were creating two single-node clusters per test which were never used.
Instead, we can now create zero-node clusters and run one-shot containers on
it (one would argue that that's still weird, but it seems acceptable for the
time being).

The newly added reference test will crash when run for commits between #7310
and #7429. See cockroachdb/postgres-test#13 for the initial state used in the
test.

tbg committed:
storage: skip flaky TestReplicaDestroy

tschottdorf commented:
Green at last.
